data-jdbc/src/main/java/io/micronaut/data/jdbc/mapper/OracleJdbcJsonColumnReader.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2017-2023 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.jdbc.mapper;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.core.annotation.Internal;
+import io.micronaut.core.type.Argument;
+import io.micronaut.data.exceptions.DataAccessException;
+import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.runtime.mapper.ResultReader;
+import io.micronaut.data.runtime.mapper.sql.SqlJsonColumnReader;
+import io.micronaut.serde.oracle.jdbc.json.OracleJdbcJsonTextObjectMapper;
+import jakarta.inject.Singleton;
+import oracle.sql.json.OracleJsonObject;
+
+import java.sql.ResultSet;
+
+/**
+ * The Oracle JDBC json column reader.
+ *
+ * @author radovanradic
+ * @since 4.0.0
+ *
+ */
+@Singleton
+@Requires(classes = OracleJdbcJsonTextObjectMapper.class)
+@Internal
+class OracleJdbcJsonColumnReader extends SqlJsonColumnReader<ResultSet> {
+
+    /**
+     * The default constructor.
+     *
+     * @param oracleJdbcJsonTextObjectMapper the oracle JSON object mapper
+     */
+    public OracleJdbcJsonColumnReader(OracleJdbcJsonTextObjectMapper oracleJdbcJsonTextObjectMapper) {
+        super(oracleJdbcJsonTextObjectMapper);
+    }
+
+    @Override
+    public Dialect getDialect() {
+        return Dialect.ORACLE;
+    }
+
+    @Override
+    public <T> T readJsonColumn(ResultReader<ResultSet, String> resultReader, ResultSet resultSet, String columnName, Argument<T> argument) {
+        try {
+            OracleJsonObject oracleJsonObject = resultSet.getObject(columnName, OracleJsonObject.class);
+            if (oracleJsonObject == null) {
+                return null;
+            }
+            byte[] content = jsonMapper.writeValueAsBytes(oracleJsonObject);
```

Comment:
This is actually more inefficient. You need to get the `OracleJsonParser` object (see https://youtu.be/jg5d15-2K3Y?t=1266) and pass it to https://github.com/micronaut-projects/micronaut-serialization/blob/5f928b7117710aacafc02aa8e03a7253d9608469/serde-oracle-jdbc-json/src/main/java/io/micronaut/serde/oracle/jdbc/json/AbstractOracleJdbcJsonObjectMapper.java#L98

But it looks like we need to make that method public so will need a PR to Micronaut Serialization //cc @dstepanov

Reply:
Yes, I saw an example in serde tests as was not sure how to use it. Also, `OracleJdbcJsonBinaryObjectMapper` does not support Timestamp and Binary types from Oracle (maybe some other as well). Field etag in _metadata is treated as Binary, this is why currently I was using `OracleJdbcJsonTextObjectMapper` to be able to test locally.

Reply:
Seems we need some upstream PRs to Micronaut Serialization

Reply:
Yeah looks like a bug. Maybe you can save that file as a binary blob and add a test to the Micronaut serialization

Reply:
Not sure we have some good option as we need to convert OracleJsonObject to byte[] in order to convert to resulting entity. Maybe we could use this what I saw on Oracle JSON video on youtube https://youtu.be/jg5d15-2K3Y?t=2850
They use something like this: https://github.com/oracle/json-in-db/blob/master/JdbcExamples/src/main/java/emp/JSONB.java
but I might ask Oracle JSON guys what they would suggest.

Reply:
If we want to use `OracleJdbcJsonBinaryObjectMapper` it has few methods to convert input to entity/POJO
```
T readValue(@NonNull byte[] byteArray, @NonNull Argument<T> type)
T readValue(@NonNull ByteBuffer<?> byteBuffer, @NonNull Argument<T> type)
T readValue(@NonNull String string, @NonNull Argument<T> type)
T readValue(@NonNull InputStream inputStream, @NonNull Argument<T> type)
```
and to get these input types from `OracleJsonObject`, we must write it to byte[], InputStream or String. Any of that is inefficient since we read `OracleJsonObject` from JDBC result set. So I don't see what options we have with current API.